cpp/references/data_structures/list_digraph.cpp

```diff
@@ -1 +1,596 @@
+}
+
+void ListDigraph::erase_vertices(Array p_vertices) {
+	for (int i = 0; i < p_vertices.size(); i++) {
+		Ref<Vertex> vertex(p_vertices[i]);
+		if (vertex.is_null())
+			continue;
+		_graph.erase(vertex->get_vertex());
+	}
+}
+
+void ListDigraph::erase_arc(Ref<Arc> p_arc) {
+	if (p_arc.is_null()) {
+		ERR_NULL_PARAM(p_arc);
+		return;
+	}
+	_graph.erase(p_arc->get_arc());
+}
+
+void ListDigraph::erase_arcs(Array p_arcs) {
+	for (int i = 0; i < p_arcs.size(); i++) {
+		Ref<Arc> arc(p_arcs[i]);
+		if (arc.is_null())
+			continue;
+		_graph.erase(arc->get_arc());
+	}
+}
+
+void ListDigraph::contract(Ref<Vertex> p_vertex1, Ref<Vertex> p_vertex2, bool p_remove) {
+	if (p_vertex1.is_null()) {
+		ERR_NULL_PARAM(p_vertex1);
+		return;
+	}
+	if (p_vertex2.is_null()) {
+		ERR_NULL_PARAM(p_vertex2);
+		return;
+	}
+	_graph.contract(p_vertex1->get_vertex(), p_vertex2->get_vertex(), p_remove);
+}
+
+void ListDigraph::reserve_vertex(int p_amount) {
+	_graph.reserveNode(p_amount);
+}
+
+void ListDigraph::reserve_arc(int p_amount) {
+	_graph.reserveArc(p_amount);
+}
+
+Ref<Vertex> ListDigraph::get_vertex(int p_id) {
+	LemonNode vertex(_graph.nodeFromId(p_id));
+	if (!_graph.valid(vertex))
+		return Ref<Vertex>();
+
+	if ((*_vertex_map)[vertex].wrapper != NULL)
+		return Ref<Vertex>((*_vertex_map)[vertex].wrapper);
+
+	return Vertex::create(vertex, this);
+}
+
+Ref<Arc> ListDigraph::get_arc(int p_id) {
+	LemonArc arc(_graph.arcFromId(p_id));
+	if (!_graph.valid(arc))
+		return Ref<Arc>();
+
+	if ((*_arc_map)[arc].wrapper != NULL)
+		return Ref<Vertex>((*_arc_map)[arc].wrapper);
+
+	return Arc::create(arc, this);
+}
+
+/* ##### GDSCRIPT API ##### */
+
+bool ListDigraph::_iter_init() {
+	_vertex_it = lemon::ListDigraph::NodeIt(_graph);
+	return _vertex_it != lemon::INVALID;
+}
+
+bool ListDigraph::_iter_next() {
+	++_vertex_it;
+	return _vertex_it != lemon::INVALID;
+}
+
+Ref<Vertex> ListDigraph::_iter_get() {
+	if ((*_vertex_map)[_vertex_it].wrapper != NULL)
+		return Ref<Vertex>((*_vertex_map)[_vertex_it].wrapper);
+	else
+		return Vertex::create(_vertex_it, this);
+}
+
+Ref<Vertex> ListDigraph::add_vertex() {
+	LemonNode vertex(_graph.addNode());
+	Ref<Vertex> instance(Vertex::create(vertex, this));
+	return instance;
+}
+
+void ListDigraph::clear() {
+	_graph.clear();
+}
+
+Array ListDigraph::get_vertices() {
+	Array ret = Array();
+	for (lemon::ListDigraph::NodeIt it(_graph); it != lemon::INVALID; ++it) {
+		if ((*_vertex_map)[it].wrapper != NULL)
+			ret.append(Ref<Vertex>((*_vertex_map)[it].wrapper));
+		else
+			ret.append(Vertex::create(it, this));
+	}
+	return ret;
+}
+
+Array ListDigraph::get_arcs() {
+	Array ret = Array();
+	for (lemon::ListDigraph::ArcIt it(_graph); it != lemon::INVALID; ++it) {
+		if ((*_arc_map)[it].wrapper != NULL)
+			ret.append(Ref<Arc>((*_arc_map)[it].wrapper));
+		else
+			ret.append(Arc::create(it, this));
+	}
+	return ret;
+}
+
+Array ListDigraph::_add_vertices(Variant p_data, Variant p_return) {
```

Comment:
I think you know what my comment is for the no less than *17* nearly identical if statement and return checks happening in the following several methods. Bundle everything up into just a ERR_FAIL_PARAM and ERR_FAIL_PARAM_V pair of macros: encompass both the if statements and the return values.